src/NNCache.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+    This file is part of Leela Zero.
+    Copyright (C) 2017 Gian-Carlo Pascutto
+
+    Leela Zero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Leela Zero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Leela Zero.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "NNCache.h"
+
+#include <functional>
+
+#include "Utils.h"
+
+NNCache::NNCache(int size) : m_size(size) {}
+
+NNCache* NNCache::get_NNCache(void) {
+  static NNCache* cache = new NNCache();
+  return cache;
+}
+
+template <class T>
+inline size_t hash_combine(size_t seed, const T& v)
+{
+      std::hash<T> hasher;
+      return seed ^ (hasher(v) + 0x9e3779b9 + (seed<<6) + (seed>>2));
+}
+
+static size_t compute_hash(const Network::NNPlanes& features) {
+    size_t hash = 0;
+    for (const auto& p : features) {
+        hash = hash_combine(hash, p);
+    }
+    return hash;
+}
+
+// Return true if the feature planes are the same.
+static bool compare(const Network::NNPlanes& a, const Network::NNPlanes& b) {
+  if (a.size() != b.size()) return false;
+  for (size_t i = 0; i < a.size(); ++i) {
```

Comment:
Auto i

Reply:
I think the second loop can probably be replaced by something out of <algorithm>.

Reply:
doh. std::vector already has an operator== overload.